src/emulator/auth/operations.ts

```diff
@@ -2830,6 +2920,254 @@ function updateTenant(
+        ) &&
+        Object.prototype.hasOwnProperty.call(userRecord, field)
+      ) {
+        switch (field) {
+          case BlockingFunctionModifiableFields.CUSTOM_CLAIMS:
+            validateSerializedCustomClaims(userRecord[field]!);
+            user.customAttributes = userRecord[field];
+            break;
+          // Session claims are only returned in beforeSignIn. For more info,
+          // see https://cloud.google.com/identity-platform/docs/blocking-functions#modifying_a_user
+          case BlockingFunctionModifiableFields.SESSION_CLAIMS:
+            try {
+              extraClaims = JSON.parse(userRecord[field]!);
+            } catch {
+              throw new BadRequestError(
+                "BLOCKING_FUNCTION_ERROR_RESPONSE: ((Response has malformed session claims.))"
+              );
+            }
+            break;
+          default:
+            (user as any)[field] = (userRecord as any)[field];
+            break;
+        }
+      }
+    }
+  }
+  return { user, extraClaims };
+}
+
+function generateBlockingFunctionJwt(
+  state: ProjectState,
+  event: BlockingFunctionEvents,
+  url: string,
+  timeoutMs: number,
+  user: Partial<UserInfo>,
+  oauthTokens: {
+    oauthIdToken?: string;
+    oauthAccessToken?: string;
+    oauthRefreshToken?: string;
+    oauthTokenSecret?: string;
+    oauthExpiresIn?: string;
+  }
+): string {
+  const issuedAt = toUnixTimestamp(new Date());
+  const jwt: BlockingFunctionsJwtPayload = {
+    iss: `https://securetoken.google.com/${state.projectId}`,
+    aud: url,
+    iat: issuedAt,
+    exp: issuedAt + timeoutMs / 100,
+    event_id: randomBase64UrlStr(16),
+    event_type: event,
+    user_agent: "", // TODO: switch to express.js to get UserAgent
+    ip_address: "", // TODO: switch to express.js to get IP address
+    locale: "en",
+    user_record: {
+      uid: user.localId,
+      email: user.email,
+      email_verified: user.emailVerified,
+      display_name: user.displayName,
+      photo_url: user.photoUrl,
+      disabled: user.disabled,
+      phone_number: user.phoneNumber,
+      custom_claims: user.customAttributes,
+    },
+    sub: user.localId,
+  };
+
+  if (state instanceof TenantProjectState) {
+    jwt.tenant_id = state.tenantId;
+    jwt.user_record.tenant_id = state.tenantId;
+  }
+
+  const provider_data = [];
+  if (user.providerUserInfo) {
+    for (const providerUserInfo of user.providerUserInfo) {
+      const provider: Provider = {
+        provider_id: providerUserInfo.providerId,
+        display_name: providerUserInfo.displayName,
+        photo_url: providerUserInfo.photoUrl,
+        email: providerUserInfo.email,
+        uid: providerUserInfo.providerId,
+      };
+      provider_data.push(provider);
+    }
+  }
+  if (user.localId) {
+    const allPhoneNumbers = state.getAllPhoneNumbersByLocalId(user.localId);
+    for (const phoneNumber of allPhoneNumbers) {
+      const provider: Provider = {
```

Comment:
I think the Auth Emulator implementation is funkier than the internal link -- since the Auth Emulator tries to keep `user.providerUserInfo` synced with phone numbers when writing to DB, Line 3191 will transform that entry into `provider_data`. The logic here will probably add yet another entry to `provider_data`, creating a duplicate. Feel free to add a test to prove me wrong.